polytracker/taint_dag.py

```diff
@@ -0,0 +1,360 @@
+        return f"affects control flow {self.affects_control_flow}"
+
+
+class TDSourceNode(TDNode):
+    def __init__(self, idx: int, offset: int, affects_control_flow: bool = False):
+        super().__init__(affects_control_flow)
+        self.idx = idx
+        self.offset = offset
+
+    def __repr__(self) -> str:
+        return f"TDSourceNode: {super().__repr__()} idx {self.idx} offset {self.offset}"
+
+
+class TDRangeNode(TDNode):
+    def __init__(self, first: int, last: int, affects_control_flow: bool = False):
+        super().__init__(affects_control_flow)
+        # First label of the range
+        self.first = first
+        # Last label of the range
+        self.last = last
+
+    def __repr__(self) -> str:
+        return f"TDRangeNode: {super().__repr__()} [{self.first}, {self.last}]"
+
+
+class TDUnionNode(TDNode):
+    def __init__(self, left: int, right: int, affects_control_flow: bool = False):
+        super().__init__(affects_control_flow)
+        self.left = left
+        self.right = right
+
+    def __repr__(self) -> str:
+        return f"TDUnionNode: {super().__repr__()} ({self.left}, {self.right})"
+
+
+class TDSink(Structure):
+    _pack_ = 1
+    _fields_ = [("fdidx", c_uint8), ("offset", c_uint64), ("label", c_uint32)]
+
+    def __repr__(self) -> str:
+        return f"TDSink fdidx: {self.fdidx} offset: {self.offset} label: {self.label}"
+
+
+class TDFile:
+    def __init__(self, buffer: bytearray) -> None:
+        # This needs to be kept in sync with implementation in encoding.cpp
+        self.source_taint_bit_shift = 63
+        self.affects_control_flow_bit_shift = 62
+        self.label_bits = 31
+        self.label_mask = 0x7FFFFFFF
+        self.val1_shift = self.label_bits
+        self.source_index_mask = 0xFF
+        self.source_index_bits = 8
+        self.source_offset_mask = (1 << 54) - 1
+
+        self.buffer = buffer
+        self.header = TDHeader.from_buffer_copy(self.buffer)
+
+        self.raw_nodes: Dict[int, int] = defaultdict(int)
+        self.sink_cache: Dict[int, TDSink] = defaultdict(TDSink)
+
+        self.fd_headers: List[Tuple[str, TDFDHeader]] = list(self.read_fd_headers())
+
+    def read_fd_headers(self) -> Iterator[Tuple[str, TDFDHeader]]:
+        assert self.header.fd_mapping_offset > 0
+        assert self.header.fd_mapping_size > 0
+
+        offset = self.header.fd_mapping_offset
+        end = offset + self.header.fd_mapping_size
+
+        while offset < end:
+            fdhdr = TDFDHeader.from_buffer_copy(self.buffer, offset)
+            offset += sizeof(TDFDHeader)
+            path = str(self.buffer[offset : offset + fdhdr.namelen], "utf-8")
+            yield (path, fdhdr)
+            offset += len(path)
+
+    @property
+    def label_count(self):
+        return int(self.header.tdag_mapping_size / sizeof(c_uint64))
+
+    def read_node(self, label: int) -> int:
+        if label in self.raw_nodes:
+            return self.raw_nodes[int]
```

Comment:
Any speedup from this caching? I would assume it would make things slower since the results is more or less just a read from the mmap-region? But, as is common with optimizations it is hard to guess...

Reply:
Probably not to be honest. I'll see if it makes any difference once I have cavity detection working and remove if not.